go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy_namespace.go

```diff
@@ -33,6 +39,148 @@ func (c *Controller) processNextANPNamespaceWorkItem(wg *sync.WaitGroup) bool {
 	return true
 }
 
+// syncAdminNetworkPolicyNamespace decides the main logic everytime
+// we dequeue a key from the anpNamespaceQueue cache
 func (c *Controller) syncAdminNetworkPolicyNamespace(key string) error {
+	// TODO(tssurya): This global lock will be inefficient, we will do perf runs and improve if needed
+	c.Lock()
+	defer c.Unlock()
+	startTime := time.Now()
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof("Processing sync for Namespace %s in Admin Network Policy controller", name)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing Namespace %s Admin Network Policy controller: took %v", name, time.Since(startTime))
+	}()
+	namespace, err := c.anpNamespaceLister.Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// (i) namespace add
+	// (ii) namespace update because namespace's labels changed
+	// (iii) namespace delete
+	// In all these cases check which ANPs were managing this namespace and enqueue them back to anpQueue
+	existingANPs, err := c.anpLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	// case (iii)
+	if namespace == nil {
+		for _, anp := range existingANPs {
+			anpObj, loaded := c.anpCache[anp.Name]
+			if !loaded {
+				continue
+			}
+			c.clearNamespaceForANP(name, anpObj)
+		}
+		return nil
+	}
+	// case (i)/(ii)
+	for _, anp := range existingANPs {
+		anpObj, loaded := c.anpCache[anp.Name]
+		if !loaded {
+			continue
+		}
+		c.setNamespaceForANP(namespace, anpObj)
+	}
```

Comment:
yeah excellent question, I had to think for a second why I used the cache instead of API, then remembered:

1. user can create two ANP's at priority 5, this is not supported by OVNK since the scenario is undefined, so only first ANP get's created, second one stays in kapi but not contained in cache, so we don't care about that one for reconciling pods
2. user can create an ANP at > 99th priority which is also not supported, same as above

So in these cases, we don't add these to the cache versus they could be in the kapi

Reply:
oh and also it could be that an API add just came in as we are looping this list but the ANP object is not yet created because c.Lock is held by `syncAdminNetworkPolicyNamespace` ; so if the setup is not done for that then probably nothing to update for the pod yet, the ANP add that will get processed next will take care of adding the pod as well..
so using cache guarantees we are looping over ANPs for which setup is done, not just ones that exist in API.

Reply:
> 1. user can create two ANP's at priority 5, this is not supported by OVNK since the scenario is undefined, so only first ANP get's created, second one stays in kapi but not contained in cache, so we don't care about that one for reconciling pods

How do we handle this on startup, where we don't have a cache? If we have two ANPs at priority 5, could we flap from one to the other on startup?

> 2. user can create an ANP at > 99th priority which is also not supported, same as above

But this is just just ignoring the ANP, nothing to do wit the cache?

> oh and also it could be that an API add just came in as we are looping this list but the ANP object is not yet created because c.Lock is held by `syncAdminNetworkPolicyNamespace` ; so if the setup is not done for that then probably nothing to update for the pod yet, the ANP add that will get processed next will take care of adding the pod as well.. so using cache guarantees we are looping over ANPs for which setup is done, not just ones that exist in API.

You would evaluate the ANPs from the informer instead and queue the needed ones. No actual change if you re-queue something that is already queued.

Reply:
Any, for the priorities we already have the `anpPriorityMap map[int32]string`, we wouldn't need the `anpCache map[string]*adminNetworkPolicyState` for that?

Reply:
> > 1. user can create two ANP's at priority 5, this is not supported by OVNK since the scenario is undefined, so only first ANP get's created, second one stays in kapi but not contained in cache, so we don't care about that one for reconciling pods
> 
> How do we handle this on startup, where we don't have a cache? If we have two ANPs at priority 5, could we flap from one to the other on startup?

oh ! this is something that slipped my mind.. you are right, we will flap and that is not what we should do. I will make sure I add logic to my repair that can deal with this. Ideally speaking I should do a sync func there for pre-populating my `anpPriorityMap` -> I know which one was the one that was setup and which one wasn't setup based on status conditions... OR creation timestamp. good catch @jcaamano !

> 
> > 2. user can create an ANP at > 99th priority which is also not supported, same as above
> 
> But this is just just ignoring the ANP, nothing to do wit the cache?

right so i could just ignore the >P99 and go on with using API, but just thought the cache guarantees I am matching the pods against ANP's whose setup is done.

> 
> > oh and also it could be that an API add just came in as we are looping this list but the ANP object is not yet created because c.Lock is held by `syncAdminNetworkPolicyNamespace` ; so if the setup is not done for that then probably nothing to update for the pod yet, the ANP add that will get processed next will take care of adding the pod as well.. so using cache guarantees we are looping over ANPs for which setup is done, not just ones that exist in API.
> 
> You would evaluate the ANPs from the informer instead and queue the needed ones. No actual change if you re-queue something that is already queued.

hmm also true, since its level, eventually everything will just reconcile, even if i added the duplicated priority ACL or the > 99th priority ACL it is going to be no-op at the end of the day, but one way or another either the ANP stuff will be handled via onANPUpdate path or via this path. I don't see a specific argument for kapi v/s cache because at the end of the day we do fetch the fresh kapi value in `syncAdminNetworkPolicy`.. this is just the key check

Reply:
> Anyway, for the priorities we already have the `anpPriorityMap map[int32]string`, we wouldn't need the `anpCache map[string]*adminNetworkPolicyState` for that?

yeah we don't need to use anpCache for priority check, its just that we store the pods/namespaces cache inside `adminNetworkPolicyState` and so its just easy to use that and pass it into set/clear functions.

Reply:
> > How do we handle this on startup, where we don't have a cache? If we have two ANPs at priority 5, could we flap from one to the other on startup?
> 
> oh ! this is something that slipped my mind.. you are right, we will flap and that is not what we should do. I will make sure I add logic to my repair that can deal with this. Ideally speaking I should do a sync func there for pre-populating my `anpPriorityMap` -> I know which one was the one that was setup and which one wasn't setup based on status conditions... OR creation timestamp. good catch @jcaamano !

I fixed this in repair.